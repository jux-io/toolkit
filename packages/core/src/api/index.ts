export * from './JuxAPI';
export * from './OrganizationsAPI';
