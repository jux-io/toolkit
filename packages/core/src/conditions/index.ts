export * from './conditions-manager';
export * from './parse-screens';
