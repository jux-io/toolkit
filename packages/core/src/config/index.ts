export * from './config.types';
export * from './find-config';
export * from './load-config';
export * from './setup-config';
export * from './validate-config';
export * from './jux-context';
export * from './get-and-verify-internal-config';
