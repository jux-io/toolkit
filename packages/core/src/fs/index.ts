export * from './fs.ts';
export * from './fs.utils.ts';
