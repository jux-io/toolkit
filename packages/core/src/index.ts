export * from './api';
export * from './assets';
export * from './config';
export * from './fs';
export * from './tokens';
export * from './utils';
export * from './auth';
export * from './postcss';
export * from './conditions';
export * from './utilities';

export { AxiosError } from 'axios';
