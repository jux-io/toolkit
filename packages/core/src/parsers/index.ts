export * from './evaluate-object';
export * from './get-property-name';
export * from './matcher-types';
