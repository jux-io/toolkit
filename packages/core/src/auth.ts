import clc from 'cli-color';
import { errors, Issuer, TokenSet } from 'openid-client';
import open from 'open';
import * as util from 'util';
import enquirer from 'enquirer';
import { logger } from './utils';
import type { UserTokens } from './config';
import { getCliConfigEnv } from './config';
import ora from 'ora';

async function generateClient() {
  const cliConfig = getCliConfigEnv();

  // fetches the .well-known endpoint for endpoints, issuer value etc.
  const auth0 = await Issuer.discover(`https://${cliConfig.AUTH_DOMAIN}`);

  // instantiates a client
  return new auth0.Client({
    client_id: cliConfig.CLIENT_ID,
    token_endpoint_auth_method: 'none',
    id_token_signed_response_alg: 'RS256',
  });
}

export async function getNewAccessToken(
  refreshToken: string
): Promise<UserTokens> {
  const client = await generateClient();

  const tokenSet = await client.refresh(refreshToken);

  if (!tokenSet.access_token || !tokenSet.expires_at) {
    throw new Error('Token set is invalid. Could not refresh token.');
  }

  return {
    access_token: tokenSet.access_token,
    refresh_token: refreshToken,
    expires_at: tokenSet.expires_at,
  };
}

export const auth = async (): Promise<UserTokens | null> => {
  const cliConfig = getCliConfigEnv();

  const scope = ['openid', 'profile', 'email', 'offline_access'].join(' ');

  const client = await generateClient();

  // Device Authorization Request - https://tools.ietf.org/html/rfc8628#section-3.1
  const handle = await client.deviceAuthorization({
    scope,
    audience: cliConfig.API_AUDIENCE,
  });

  // Device Authorization Response - https://tools.ietf.org/html/rfc8628#section-3.2
  const { verification_uri_complete, user_code, expires_in } = handle;

  // User Interaction - https://tools.ietf.org/html/rfc8628#section-3.3
  await enquirer.prompt([
    {
      type: 'invisible',
      name: 'prompt',
      // eslint-disable-next-line max-len
      message: util.format(
        `${clc.underline(
          'Press any key'
        )} to open up the browser to login or press ctrl-c to abort. You should see the following code: ${clc.green.bgWhite.underline(
          user_code
        )}. It expires in ${
          expires_in % 60 === 0
            ? `${expires_in / 60} minutes`
            : `${expires_in} seconds`
        }.`
      ),
    },
  ]);

  // opens the verification_uri_complete URL using the system-register handler for web links (browser)
  open(verification_uri_complete);

  const spinner = ora(`Waiting for confirmation...`).start();

  // Device Access Token Request - https://tools.ietf.org/html/rfc8628#section-3.4
  // Device Access Token Response - https://tools.ietf.org/html/rfc8628#section-3.5
  let tokens: TokenSet;
  try {
    tokens = await handle.poll();
    spinner.succeed('Done');
  } catch (err) {
    switch (err.error) {
      case 'access_denied': // end-user declined the device confirmation prompt, consent or rules failed
        // logger.warn(`Confirmation was not granted.`);
        spinner.fail('Confirmation was not granted. Aborting.');
        break;
      case 'expired_token': // end-user did not complete the interaction in time
        logger.warn(`Token has expired, you need to login again.`);
        spinner.warn('Token has expired, you need to login again.');
        break;
      default:
        if (err instanceof errors.OPError) {
          logger.error(
            util.format(
              `Login failed. error = ${err.error}; error_description = ${err.error_description}`
            )
          );
          spinner.fail(
            `Login failed. Error description: ${err.error_description}`
          );
        } else {
          throw err;
        }
    }

    return null;
  }

  if (!tokens.access_token || !tokens.refresh_token || !tokens.expires_at) {
    throw new Error('Token set is invalid. Could not refresh token.');
  }

  return {
    access_token: tokens.access_token as string,
    refresh_token: tokens.refresh_token as string,
    expires_at: tokens.expires_at as number,
  };
};
