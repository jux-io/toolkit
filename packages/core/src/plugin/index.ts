import type { UnpluginFactory } from 'unplugin';
import { createUnplugin } from 'unplugin';
import { cosmiconfigSync } from 'cosmiconfig';
import type { Options } from './types';

import path from 'path';
import { logger } from '../utils';
import { JuxCLIConfig } from '../config';

export const unpluginFactory: UnpluginFactory<Options | undefined> = (
  options
) => {
  const explorer = cosmiconfigSync('jux');
  const explorerRes = explorer.search(process.cwd() + '/astro-app');

  if (!explorerRes) {
    logger.error(`Config not found at ${process.cwd()}`);
  }

  let config: JuxCLIConfig = explorerRes?.config;

  return {
    name: '@juxio/core',
    enforce: 'post',
    transformInclude(id) {
      return !config || !id.includes('node_modules');
    },
    watchChange(id /*,change*/) {
      if (id === explorerRes?.filepath) {
        // Reload the config
        explorer.clearCaches();
        const newExplorerRes = explorer.search(process.cwd() + '/astro-app');
        config = newExplorerRes?.config;
      }
    },
    transform(code: string /*,url: string*/) {
      this.addWatchFile(path.join(process.cwd(), 'astro-app', 'jux.config.ts'));
      // const [id] = url.split('?', 1);

      return code.replace('__UNPLUGIN__', `Hello Unplugin! ${options}`);
    },
  };
};

export const unplugin = /* #__PURE__ */ createUnplugin(unpluginFactory);

export default unplugin;
