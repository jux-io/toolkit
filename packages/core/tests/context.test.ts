import { describe, it, expect } from 'vitest';
import { getConfigContext } from '../src';
import { join, resolve } from 'node:path';

const root = resolve(join(__dirname, '../'));

const cwd = join(root, 'tests', 'sample-config');

describe('Context', () => {
  it('Context loads the right styling configurations', async () => {
    const ctx = await getConfigContext({
      cwd,
    });

    await ctx.stylesheetManager.appendBaseStyles();

    expect(ctx.stylesheetManager.toCss()).toMatchSnapshot();
  });
});
