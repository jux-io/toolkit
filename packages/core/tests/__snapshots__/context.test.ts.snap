// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Context > Context loads the right styling configurations 1`] = `
"@layer juxbase {*, ::before, ::after, ::backdrop, ::file-selector-button {
  box-sizing: border-box;
  border-width: 0;
  border-style: solid;
  border-color: var(--global-color-border, currentColor);
}html, :host {
  line-height: 1.5;
  -webkit-text-size-adjust: 100%;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  -moz-tab-size: 4;
  tab-size: 4;
  font-family: var(--global-font-body, "ui-sans-serif, system-ui, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, 'Helvetica Neue', Arial, 'Noto Sans', sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji'");
  font-feature-settings: var(--global-font-feature-settings, normal);
  font-variation-settings: var(--global-font-variation-settings, normal);
  -webkit-tap-highlight-color: transparent;
}body {
  margin: 0;
  line-height: inherit;
}hr {
  height: 0;
  color: inherit;
  border-top-width: 1px;
}abbr:where([title]) {
  text-decoration: underline dotted;
}h1, h2, h3, h4, h5, h6 {
  font-size: inherit;
  font-weight: inherit;
}a {
  color: inherit;
  text-decoration: inherit;
}b, strong {
  font-weight: bolder;
}code, kbd, samp, pre {
  font-family: var(--global-font-mono, "ui-monospace, SFMono-Regular, Menlo, Monaco, Consolas, 'Liberation Mono', 'Courier New'");
  font-feature-settings: var(--global-font-feature-settings, 'normal');
  font-variation-settings: var(--global-font-variation-settings, 'normal');
  font-size: 1em;
}small {
  font-size: 80%;
}sub, sup {
  font-size: 75%;
  line-height: 0;
  position: relative;
  vertical-align: baseline;
}sub {
  bottom: -0.25em;
}sup {
  top: -0.5em;
}table {
  text-indent: 0;
  border-color: inherit;
  border-collapse: collapse;
}button, input, optgroup, select, textarea, ::file-selector-button {
  font: inherit;
  font-feature-settings: inherit;
  font-variation-settings: inherit;
  letter-spacing: inherit;
  color: inherit;
  background: transparent;
}button, select {
  text-transform: none;
}button, input:where([type="button"]), input:where([type="reset"]), input:where([type="submit"]), ::file-selector-button {
  -webkit-appearance: button;
  background-color: transparent;
  background-image: none;
}:-moz-focusring {
  outline: auto;
}:-moz-ui-invalid {
  box-shadow: none;
}progress {
  vertical-align: baseline;
}::-webkit-inner-spin-button, ::-webkit-outer-spin-button {
  height: auto;
}[type="search"] {
  -webkit-appearance: textfield;
  outline-offset: -2px;
}::-webkit-search-decoration, ::-webkit-search-cancel-button {
  -webkit-appearance: none;
}::-webkit-file-upload-button {
  -webkit-appearance: button;
  font: inherit;
}summary {
  display: list-item;
}blockquote, dl, dd, h1, h2, h3, h4, h5, h6, hr, figure, p, pre {
  margin: 0;
}fieldset {
  margin: 0;
  padding: 0;
}legend {
  padding: 0;
}ol, ul, menu {
  list-style: none;
  margin: 0;
  padding: 0;
}dialog {
  padding: 0;
}textarea {
  resize: vertical;
}input::placeholder, textarea::placeholder {
  color: var(--global-color-placeholder, 'color-mix(in srgb, currentColor 50%, transparent)');
  opacity: 1;
}:disabled {
  cursor: default;
}img {
  border-style: none;
}img, svg, video, canvas, audio, iframe, embed, object {
  display: block;
  vertical-align: middle;
}img, video {
  max-width: 100%;
  height: auto;
}[hidden] {
  display: none;
}* {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
  background-color: var(--core-color-brand_100);
  color: var(--does_not_exist);
  margin-top: var(--some_weird_token_with_special_chars);
  top: var(--wow);
}html, body {
  font-family: Inter;
  font-size: 16px;
  color: #111827;
  background-color: #FFF;
}
}
@layer juxtokens {:where(:root, :host) {--core-color-brand_100: #0070f3;
--core-color-brand_200: #ff0080;
--core-typography-12_bold-font_size: var(--core-dimension-spacing_100);
--core-typography-12_bold-font_family: Inter;
--core-typography-12_bold-font_weight: 700;
--core-typography-12_bold-line_height: 16px;
--core-typography-12_bold-letter_spacing: 12px;
--core-dimension-spacing_100: 8px;
}[data-jux-theme="light"] {--color-primary: var(--core-color-brand_100);
--color-secondary: #FFF;
--dimension-specific-button_radius: var(--core-dimension-spacing_100);
}[data-jux-theme="dark"] {--color-primary: var(--core-color-brand_200);
--color-secondary: var(--color-primary);
--dimension-specific-button_radius: var(--core-dimension-spacing_100);
}
}
@layer juxutilities {}"
`;
